### PR TITLE
feat: add `ChangeColumnType` for `AlterKind`

### PR DESCRIPTION
**Pull request by KKould**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
https://github.com/GreptimeTeam/greptimedb/issues/3517

## What's changed and what's your intention?

add `ChangeColumnType` to `AlterKind` and validate `ChangeColumnType`, process the meta after `ChangeColumnType`

Tips: The `AlterKind::ChangeColumnType` part of mapping greptime-proto will be committed later (after the update of greptime-proto)

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by evenyag** [comment]
I'll review this today if possible.

File: src/table/src/metadata.rs [hunk 1]
```diff
@@ -458,6 +459,131 @@ impl TableMeta {
...
+                    }
+                );
+
+                ensure!(
+                    table_schema.contains_column(&col_to_modify.column_name),
+                    error::ColumnExistsSnafu {
+                        table_name,
+                        column_name: col_to_modify.column_name.to_string()
+                    },
+                );
```
Comment (evenyag) [inline src/table/src/metadata.rs:514]: We already get column by name, do we need to check this again?

File: src/table/src/metadata.rs [hunk 2]
```diff
@@ -458,6 +459,131 @@ impl TableMeta {
...
 
+    fn modify_columns(
+        &self,
+        table_name: &str,
+        requests: &[ModifyColumnRequest],
+    ) -> Result<TableMetaBuilder> {
+        let table_schema = &self.schema;
+        let mut meta_builder = self.new_meta_builder();
+
+        let mut modify_columns: HashMap<_, _> = HashMap::with_capacity(requests.len());
```
Comment (evenyag) [inline src/table/src/metadata.rs:470]:
```suggestion
        let mut modify_columns = HashMap::with_capacity(requests.len());
```

File: src/table/src/metadata.rs [hunk 3]
```diff
@@ -458,6 +459,131 @@ impl TableMeta {
...
+        let timestamp_index = table_schema.timestamp_index();
+
+        for col_to_modify in requests {
+            let modify_column_name = &col_to_modify.column_name;
+
+            if let Some(index) = table_schema.column_index_by_name(modify_column_name) {
```
Comment (evenyag) [inline src/table/src/metadata.rs:476]: Use `OptionExt::context` or `OptionExt::with_context`
https://docs.rs/snafu/latest/snafu/trait.OptionExt.html#tymethod.context

File: src/store-api/src/region_request.rs [hunk 4]
```diff
@@ -501,6 +515,60 @@ impl TryFrom<v1::AddColumnLocation> for AddColumnLocation {
...
+impl ModifyColumn {
+    /// Returns an error if the column to modify is invalid.
+    pub fn validate(&self, metadata: &RegionMetadata) -> Result<()> {
+        let column_meta =
+            metadata
+                .column_by_name(&self.column_name)
+                .context(InvalidRegionRequestSnafu {
+                    region_id: metadata.region_id,
+                    err: format!("column {} not found", self.column_name),
+                })?;
```
Comment (evenyag) [inline src/store-api/src/region_request.rs:536]: Avoid constructing the struct as it uses `format`.
```suggestion
                .with_context(|| InvalidRegionRequestSnafu {
                    region_id: metadata.region_id,
                    err: format!("column {} not found", self.column_name),
                })?;
```

File: src/store-api/src/region_request.rs [hunk 5]
```diff
@@ -501,6 +515,60 @@ impl TryFrom<v1::AddColumnLocation> for AddColumnLocation {
...
+
+        ensure!(
+            !matches!(
+                column_meta.semantic_type,
+                SemanticType::Timestamp | SemanticType::Tag
+            ),
```
Comment (evenyag) [inline src/store-api/src/region_request.rs:542]:
```suggestion
            matches!(
                column_meta.semantic_type,
                SemanticType::Field
            ),
```

File: src/store-api/src/region_request.rs [hunk 6]
```diff
@@ -501,6 +515,60 @@ impl TryFrom<v1::AddColumnLocation> for AddColumnLocation {
...
+                    "column '{}' cannot be cast automatically to type '{}'",
+                    self.column_name, self.target_type
+                ),
+            }
+        );
+
+        Ok(())
+    }
+
+    /// Returns true if no column to modify to the region.
```
Comment (killme2008) [inline src/store-api/src/region_request.rs:565]: Looks like the comment is incorrect?

**Review by killme2008 — COMMENTED** [review]
If the purpose of `ModifyColumn` is solely to alter the column type, then `ChangeType` would be a more apt name. @evenyag What do you think?

File: src/store-api/src/region_request.rs (hunk 6 above)
Comment (KKould) [reply src/store-api/src/region_request.rs:565]: > Looks like the comment is incorrect?

yep, actually I have reported this to @waynexia before, and I intentionally kept this issue in order to maintain uniformity.

**Comment by KKould** [comment]
> If the purpose of `ModifyColumn` is solely to alter the column type, then `ChangeType` would be a more apt name. @evenyag What do you think?

I plan to implement renaming column name after this pr
issue: https://github.com/GreptimeTeam/greptimedb/issues/996

**Comment by killme2008** [comment]
> > If the purpose of `ModifyColumn` is solely to alter the column type, then `ChangeType` would be a more apt name. @evenyag What do you think?
> 
> I plan to implement renaming column name after this pr issue: #996

I prefer that each `AlterKind` only performs a single function, in line with the Single Responsibility Principle (SRP: https://en.wikipedia.org/wiki/Single-responsibility_principle). Therefore, I personally recommend using a different `AlterKind` for renaming columns.

**Comment by KKould** [comment]
> > > If the purpose of `ModifyColumn` is solely to alter the column type, then `ChangeType` would be a more apt name. @evenyag What do you think?
> > 
> > 
> > I plan to implement renaming column name after this pr issue: #996
> 
> I prefer that each `AlterKind` only performs a single function, in line with the Single Responsibility Principle (SRP: https://en.wikipedia.org/wiki/Single-responsibility_principle). Therefore, I personally recommend using a different `AlterKind` for renaming columns.

agree, then I will make changes later

**Review by killme2008 — APPROVED** [review]
LGTM. Great work! Don't forget to change the PR title too.

File: src/store-api/src/metadata.rs [hunk 7]
```diff
@@ -535,6 +537,7 @@ impl RegionMetadataBuilder {
         match kind {
             AlterKind::AddColumns { columns } => self.add_columns(columns)?,
             AlterKind::DropColumns { names } => self.drop_columns(&names),
+            AlterKind::ChangeColumnTypes { columns } => self.change_columns_type(columns),
```
Comment (evenyag) [inline src/store-api/src/metadata.rs:540]: Keep consistency.
```suggestion
            AlterKind::ChangeColumnTypes { columns } => self.change_column_types(columns),
```

File: src/store-api/src/metadata.rs [hunk 8]
```diff
@@ -615,6 +618,25 @@ impl RegionMetadataBuilder {
         self.column_metadatas
             .retain(|col| !name_set.contains(&col.column_schema.name));
     }
+
+    /// Changes columns type to the metadata if exist.
+    fn change_columns_type(&mut self, columns: Vec<ChangeColumnType>) {
```
Comment (evenyag) [inline src/store-api/src/metadata.rs:623]:
```suggestion
    fn change_column_types(&mut self, columns: Vec<ChangeColumnType>) {
```

**Pull request merged**